src/createEventEmitterProxy.ts

```diff
@@ -93,8 +107,40 @@ export function createEventEmitterProxy<Type extends EventEmitterLike>(
       }
 
       const value = target[name];
-      if (value instanceof Function) {
+
+      if (typeof value === 'function') {
         return function (this: unknown, ...args: any[]) {
+          if (name === 'once') {
+            const unwrappedHandler = args[1];
+            const wrappedHandler = (...handlerArgs: any[]) => {
+              removeEvent(args[0], wrappedHandler);
+              return unwrappedHandler(...handlerArgs);
+            };
+            args[1] = wrappedHandler;
+          }
+          if (
+            name === 'on' ||
+            name === 'addListener' ||
+            name === 'prependListener' ||
+            name === 'once'
+          ) {
+            if (eventFilter(args[0])) {
+              eventsAdded.push({
+                addedWith: name,
+                name: args[0],
+                handler: args[1],
+              });
+            } else {
+              return target;
```

Comment:
Hmm, do we want to return the `target` here or return the `value`? It seems like we need tests to handle that an event that's being filtered out can still be listened to via `on`, `addListener`, `prependListener`, and `once`.

``` suggestion
              // This function may be either bound to something or nothing.
              // eslint-disable-next-line no-invalid-this
              return value.apply(this === receiver ? target : this, args);

Reply:
Perhaps easiest to drop this block, and use `else if` for the `off`/`removeListener` check. Then it always falls through to the final `value.apply(this === receiver ? target : this, args);`

Reply:
Hmm maybe I've misunderstood the purpose of the eventFilter. I was under the impression that when we filter an event, we are not wanting to events of that type. Given that, why would we want to add the subscription if the event is supposed to be filtered out?

The reason I return target here is because the return value when calling `on` (or equivalent) is always the event emitter itself (which is `target` here).

Reply:
`eventFilter` shouldn't suppress any events. The filtered events are excluded from _migration_, i.e. they don't get migrated when the target changes.

The code looks correct in that respect though. You're using the event filter correctly here I think.

>The reason I return target here is because the return value when calling on (or equivalent) is always the event emitter itself (which is target here).

This is the getter though, so you'd be returning the event emitter itself for `eventEmitter.on`. We'd want the `on` function to be returned instead, no? Maybe I'm missing something.

Reply:
Ah thank you for clarifying. In this case, I believe my implementation is incorrect. Specifically, consider this code:
```
       if (
            name === 'on' ||
            name === 'addListener' ||
            name === 'prependListener' ||
            name === 'once'
          ) {
            if (eventFilter(args[0])) {
              eventsAdded.push({
                addedWith: name,
                name: args[0],
                handler: args[1],
              });
            } else {
              return target;
            }
          }
          if (name === 'off' || name === 'removeListener') {
            removeEvent(args[0], args[1]);
          }

          // This function may be either bound to something or nothing.
          // eslint-disable-next-line no-invalid-this
          return value.apply(this === receiver ? target : this, args);
        };
```

Assume the following:
 1. we call `proxy.on('foo', () => true)`. 
 2. we have eventFilter list ['foo']
 
 With our code, `value.apply(...)` would not be called, and thus the subscription would never be made.
 
 According to what you've explained here, this is incorrect. It should still create the subscription, it just should not migrate the subscription when the target is changed.

Reply:
Right, exactly. And this would be fixed by the suggestions Elliot and I left in the first two comments, I think.